tests/trunner.nim

```diff
@@ -1,124 +1,63 @@
-      let testPath = prepareFiles(conf, tmpDir)
-      discard run(testPath)
-      let xmlPath = conf.outputDir / "results.xml"
-      let xmlContents = readFile(xmlPath)
-      check:
-        xmlContents.len > 0
-        "failure" notin xmlContents
-      moveFile(xmlPath, conf.outputDir / slugUnder & ".xml")
+for status in ["pass", "fail", "error"]:
+  for (kind, path) in walkDir(getAppDir() / status):
+    if kind == pcDir:
+      for file in walkFiles(path / "*_test.nim"):
+        suite status & '/' & path.splitPath().tail:
+          let slugUnder = file.splitFile().name[0..^6]
+          let slug = slugUnder.replace('_', '-')
+          let conf = Conf(slug: slug,
+                          inputDir: path & '/',
+                          outputDir: outputDir)
+          removeDir(conf.outputDir)
+          createDir(conf.outputDir)
+
+          let tmpDir = createTmpDir()
+          let testPath = prepareFiles(conf, tmpDir)
+
+          test "prepareFiles: Copies the input solution; Returns the test path":
+            check tmpDir == tmpBase / "nim_test_runner"
+            check testPath == tmpDir / slugUnder & "_test.nim"
+            let solNameAndExt = slugUnder & ".nim"
+            let tmpSolutionContents = readFile(tmpDir / solNameAndExt)
+            let origSolutionContents = readFile(conf.inputDir / solNameAndExt)
+            check tmpSolutionContents == origSolutionContents
+
+          if slug == "hello-world" and status == "pass":
+            test "prepareFiles: The `hello_world` test file is as expected":
+              let f = readFile(conf.inputDir / "expected_hello_world_test_prepared.nim")
+              let expectedHelloWorldTest = f.replace(
+                "trunner_replaces_this_with_path_to_results_json",
+                outputDir / "results.json")
+              check readFile(testPath) == expectedHelloWorldTest
+
+          copyFile(getAppDir().parentDir / "src" / "unittest_json.nim",
+                   tmpDir / "unittest_json.nim")
+
+          let expectedExitCodeOfRunProc = if status == "pass": 0 else: 1
+          test "The `run` proc returns the expected exit code":
+            check run(testPath) == expectedExitCodeOfRunProc
+
+          let resultsJson = parseFile(conf.outputDir / "results.json")
+          let pathExpectedResultsJson = path / "expected_results.json"
+          test "The `results.json` file is as expected":
+            if existsFile(pathExpectedResultsJson):
+              let expectedResultsJson = parseFile(pathExpectedResultsJson)
+              if resultsJson != expectedResultsJson:
+                echo "\nresults.json:"
+                echo resultsJson.pretty()
+                echo "\nexpected_results.json:"
+                echo expectedResultsJson.pretty()
+                fail()
+            else:
+              when defined(writeJson):
+                writeFile(pathExpectedResultsJson, resultsJson.pretty())
+                echo "Wrote: " & pathExpectedResultsJson
```

Comment:
What is this for?
To create the JSON files for what it should look like?

Reply:
Yes. If `writeJson` is defined at compile-time then executing the resulting binary will write an `expected_results.json` where it doesn't exist. The corresponding test still fails.

It's a shortcut so that maintainers don't have to copy and "prettify" the `results.json` from the output directory as they create new test cases. I prefer that our `expected_results.json` are in "pretty" format so they are easier to inspect; the output `results.json` currently contains no newlines. However, this means we have to compare them either by parsing both, "unprettifying" the `expected_results.json`, or "prettifying" `results.json`.

I didn't want to start parsing runtime arguments in `trunner`. This way, the JSON-writing functionality is not included in the `trunner` binary unless you compile with `nim c -d:writeJson trunner.nim`. And we will not pass this flag during CI.

Reply:
Would it not be advantageous to have a CI check that makes sure that there is a golden JSON for every test file and runs the runner with the `writeJson` flag?